### PR TITLE
Closes #33 Create `derive_params_by_age()` which creates records for growth chart data by age

### LINKED ISSUE #33: New function: derive_params_growth_age - for BY AGE growth parameters
**New function:**
```
derive_params_growth_age(
  dataset, 
  sex, 
  age, 
  age_unit, 
  meta_criteria, 
  parameter, 
  set_values_to_sds, 
  set_values_to_pctl)
```
 
**Arguments:**

- dataset: input dataset
- sex, age and age_unit would define the variables to use from input dataset: e.g. sex = SEX, age = AAGECUR, age_unit = AAGECURU
- meta_criteria: metadata as defined in step 4 of MS Teams doc
- parameter: the input dataset PARAMCD for which we want growth derivations: e.g. parameter = exprs(BMI = VSTESTCD == "BMI"). Options: “BMI”, “HEIGHT”, “WEIGHT”, “HEAD CIRCUMFERENCE”
- set_values_to_sds: variables to be set for zscore (SDS) parameter: e.g. set_values_to_sds(PARAMCD = “BMIASDS”, PARAM = “BMI-for-age z-score”). If left blank then parameter not derived in output dataset.
- set_values_to_pctl: variables to be set for percentile parameter: e.g. set_values_to_pctl(PARAMCD = “BMIAPCTL”, PARAM = “BMI-for-age Percentile”). If left blank then parameter not derived in output dataset.

**Checks:**

- dataset
  - The variables for sex, age and age_unit are in input dataset – age then first needs to get converted to DAYS using admiral::derive_vars_duration()
  - The PARAMCD in parameter is available in input dataset
  - the unit associated with each parameter is as expected (Height/Head CIRCUMFERENCE in cm, weight in kg)
- metadata
  - has variables for SEX, AGE, L, M, S 

**Derivations:**

image

**Example function call:** 
```
advs_wgt_age <- advs %>% 
  derive_params_growth_age(
    sex = SEX, 
    age = AAGECUR, 
    age_unit = AAGECURU, 
    meta_criteria = mymetadata,
    parameter = WGHT,
    set_values_to_sds(
      PARAMCD = “WGTASDS”,
      PARAM = “Weight-for-age Z-Score”
    ),
    set_values_to_pctl(
      PARAMCD = “WGTAPCTL”,
      PARAM = “Weight-for-age Percentile”
    )
  )
```

**NOTE from David for BMI:**
CDC BMI percentiles >95 are calculated differently than this <95th percentile.
Here are the 2 places that 'introduce' the new (2023) CDC calculations for BMIs > 95th percentile

1. https://www.cdc.gov/nccdphp/dnpao/growthcharts/resources/sas.htm - (This a page that I largely wrote).  You can see the big notification at the top of the screen and there's a section on 'Extended z-scores and percentiles' about halfway down the long webpage
2. And here's a page written by CDC people in Washington (Natl Center for Health Statistics): https://stacks.cdc.gov/view/cdc/122933

We called them extended BMIz and extended BMI percentiles, but the SAS and R programs name these calculations as 'BMIz' and 'BMIp'.  If someone wanted to use the old visions of the calculations to check their work done in years before 2023, the variables are named something like original_BMIz and original_BMIp

### PR DESCRIPTION
**Pull request by zdz2101**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiralpeds (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiralpeds site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/derive_params_growth_age.R [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+derive_params_growth_age <- function(dataset,
+                                     sex,
+                                     age,
+                                     age_unit,
+                                     meta_criteria,
+                                     parameter,
+                                     set_values_to_sds = NULL,
+                                     set_values_to_pctl = NULL) {
+  # Apply assertions to each argument to ensure each object is appropriate class
+  # need to add assertion for dataset that checks columns in sex, age, age_unit are present
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:171]: having trouble parsing out the correct code here, but its something along the lines of:

`assert_data_frame(dataset, required_vars = expr_c({{sex}}, {{age}}, {{age_unit}})`

File: R/derive_params_growth_age.R [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    dataset_final <- bind_rows(dataset, add_sds) %>%
+      select(-c(L, M, S, sex_join, ageu_join, prev_age, next_age))
+  }
+
+  if (!is_empty(set_values_to_pctl)) {
+    add_pctl <- added_records %>%
+      mutate(
+        AVAL = ((VSSTRESN / M)^L - 1) / (L * S),
+        AVAL = pnorm(AVAL) * 100,
+        # may need to add special modification for > 95th percentile
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:226]: need to read the modification for whats done at >= 95%-tile but anything needed belongs here

File: tests/testthat/test-derive_params_growth_age.R [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+## Test 1: derive_params_growth_age works ----
+test_that("derive_params_growth_age Test 1: derive_params_growth_age works", {
+  vs_data <- tibble::tribble(
+    ~USUBJID, ~SEX, ~AGECUR, ~AGEU, ~VSTESTCD, ~VSSTRESN,
+    "1001", "M", 30, "days", "WEIGHT", 10,
+    "1002", "F", 25, "months", "WEIGHT", 100,
+  )
+
+  fake_meta <- tibble::tribble(
```
Comment (zdz2101) [inline tests/testthat/test-derive_params_growth_age.R:11]: to avoid using dependencies in data I created an entirely fake metadata example that as long as the structure is correct will make the appropriate calcuation

File: R/derive_params_growth_age.R [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        !!!set_values_to_sds
+      )
+    dataset_final <- bind_rows(dataset, add_sds) %>%
+      select(-c(L, M, S, sex_join, ageu_join, prev_age, next_age))
+  }
+
+  if (!is_empty(set_values_to_pctl)) {
+    add_pctl <- added_records %>%
+      mutate(
+        AVAL = ((VSSTRESN / M)^L - 1) / (L * S),
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:224]: `VSSTRESN` is always present in ADVS but perhaps `AVAL` should be used here? is there a need to double-check that numeric value and unit associated with it is valid/compatible with the metadata?

**Review by zdz2101 — COMMENTED** [review]
This should be a decently rough draft to get you going @kabis-ops its missing just a couple parts still but I think it is close to what you expect to "works" 

@rossfarrugia can you take a look at the comments I've added as well, perhaps during my absence if people want to work on the smaller pieces left in this PR, it can be brought to the finish line by next week but otherwise it should be a good rough skeleton to get the template, other functions, off to a running start

File: R/derive_params_growth_age.R [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+#'     out_unit = "months",
+#'     trunc_out = FALSE
+#'   ) %>%
+#'   mutate(
+#'     AGECUR = ifelse(AGECUR_D >= 365.25 * 2, AGECUR_M, AGECUR_D),
+#'     AGECURU = ifelse(AGECUR_D >= 365.25 * 2, CURU_M, CURU_D)
+#'   )
+#'
+#' #' metadata is in months
+#' cdc_meta_criteria <- admiralpeds::cdc_wtage %>%
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:117]: Lines 117 - 144 act as a decent "guide" to how the metadata for the charts should be preprocessed

File: R/derive_params_growth_age.R [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+#'   derive_vars_duration(
+#'     new_var = AGECUR_M,
+#'     new_var_unit = CURU_M,
+#'     start_date = BRTHDT,
+#'     end_date = VSDT,
+#'     out_unit = "months",
+#'     trunc_out = FALSE
+#'   ) %>%
+#'   mutate(
+#'     AGECUR = ifelse(AGECUR_D >= 365.25 * 2, AGECUR_M, AGECUR_D),
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:112]: obviously here we have the 2 year old cut off to use whichever is the appropriate current age/unit, but depending on study this can be modfied as well, the function is agnostic to this

File: R/derive_params_growth_age.R [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+      ageu_join = AGEU
+    )
+
+  # Merge the dataset that contains the vs records and filter the L/M/S that fit the appropriate age
+  # To parse out the appropriate age, create [x, y) brackets using a prev_age <= AGECUR < next_age
+  added_records <- dataset %>%
+    filter(!!enexpr(parameter)) %>%
+    left_join(.,
+      processed_md,
+      by = c("sex_join", "ageu_join"),
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:203]: I decided by joining the entire metadataset by sex and age unit then filtering the correctly mapped L/M/S was more appropriate than creating some sort of weird fuzzy join or join_by statement using within() or overlaps(), reason this came up was imagine a dataset that has a

24 days old and 24.6 months old patient in the same study

we run into a couple issues:
- desired age units is not homogenous in the study
- metadata begins to have "gaps" between ages
- the 24 days old patient should have a pretty easy/direct mapping to a metadata entry appropriate for 24 days old, at least in the WHO data its per day, which is the smallest denomination anyway so thats fine but
- the 24.6 months old patient is in the range between 24.5 and 25.5 months old for the CDC data

I played around having the function only be able to map one age unit at a time, but running the function twice felt clunky, the way I've currently programmed it left a decent amount of flexibility depending how the user wants to configure their metadata

**Comment by Lina2689** [comment]
In `derive_params_growth_age` function, can we modify the code to assign PARAMN for Z-score and percentile parameters? As for now, its taking same PARAMN of the input parameter from VSTESTCD.

image

**Comment by Fanny-Gautier** [comment]
Once the function is applied, should it not keep the derived parameters only instead of all records from the input dataset ?
Filter is needed to keep only the newly derived ones:
image

**Comment by rossfarrugia** [comment]
@Lina2689 for `PARAMN` please use same approach in the template as here https://pharmaverse.github.io/admiral/articles/bds_finding.html#paramcd for consistency with admiral.

@Fanny-Gautier this is the usual convention for all admiral `derive_param_xxx()` functions, e.g. see https://pharmaverse.github.io/admiral/reference/derive_param_bmi.html. Also in our ADVS example (https://pharmaverse.github.io/admiral/articles/bds_finding.html) we keep the source record parameters as well as the newly derived rows. This gives greater transparency and traceability as you can see where the derived records came from in one place.

Let me know if you disagree though or have any other thoughts on this. We generally try and keep the package extensions as similar as possible to admiral so that users find it easy and familiar when swapping between them.

File: _pkgdown.yml [hunk 8]
```diff
@@ -25,6 +25,10 @@ navbar:
        href: reference/
 
 reference:
+- title: Derivations for Growth Chart
+  contents:
+  - has_keyword('der_growth')
```
Comment (rossfarrugia) [inline _pkgdown.yml:30]: @zdz2101 for closer consistency with admiral i would suggest something like:

```
- title: Derivations for Adding Parameters/Records
- subtitle: Vital Signs-specific
  contents:
  - has_keyword("der_prm_bds_vs")
```

**Review by kabis-ops — COMMENTED** [review]
Hi @zdz2101 
Thanks! i have started to review the function and left a few comments/questions. 
I did not test it yet, so it is just a  kind of code review. I will try to send any other comments/question related to the function test/results tomorrow, if any. 
Thank you again
cc: @rossfarrugia , you may have some opinion/input on some questions. Please feel to commet and/or tag others

File: R/derive_params_growth_age.R [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+#'   Expected values: 'days' 'months'
+#'
+#' @param meta_criteria Metadata dataset
+#'
+#'   A metadata dataset with the following expected variables:
+#'   `AGE`, `AGEU`, `SEX`, `L`, `M`, `S`
+#'
+#'   The dataset can be derived from CDC/WHO or user-defined datasets.
+#'   The CDC/WHO growth chart metadata datasets are available in the package and will
+#'   require small modifications.
```
[the remark this answers is not in the document]
Comment (zdz2101) [reply R/derive_params_growth_age.R:33]: I believe the metadata will be up to the user, it doesn't necessarily require functions from us but rather some `dplyr` work,  we can have some sort of metadata creation guide

File: R/derive_params_growth_age.R [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+#'
+#'   e.g. `parameter = VSTESTCD == "WEIGHT"`.
+#'
+#'   There is CDC/WHO metadata available for Height, Weight, BMI, and Head Circumference
+#'
+#' @param bmi_cdc_correction Extended CDC BMI-for-age correction
+#'
+#'  A logical scalar, e.g. `TRUE`/`FALSE` is expected.
+#'  The CDC handles Z-scores and percentiles above the 95th percentile in a different way,
+#'  if set to `TRUE` the CDC's correction is applied.
```
[the remark this answers is not in the document]
Comment (zdz2101) [reply R/derive_params_growth_age.R:54]: already set to FALSE

File: R/derive_params_growth_age.R [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+                                     bmi_cdc_correction = FALSE,
+                                     set_values_to_sds = NULL,
+                                     set_values_to_pctl = NULL) {
+  # Apply assertions to each argument to ensure each object is appropriate class
+  # need to add assertion for dataset that checks columns in sex, age, age_unit are present
+  assert_data_frame(meta_criteria, required_vars = exprs(SEX, AGE, AGEU, L, M, S))
+  assert_expr(enexpr(parameter))
+  assert_varval_list(set_values_to_sds, optional = TRUE)
+  assert_varval_list(set_values_to_pctl, optional = TRUE)
+
```
[the remark this answers is not in the document]
Comment (zdz2101) [reply R/derive_params_growth_age.R:183]: I think because the user may have so much freedom to put stuff in these arguments, its hard to check for and catch every case, I think as long as we write thoughtful vignettes and leave enough documentation, there is still some responsibility left to the user/programmer to properly implement and check what theyre feeding in is correct

File: R/derive_params_growth_age.R [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+  # create a unified join naming convention, hard to figure out in by argument
```
[the remark this answers is not in the document]
Comment (zdz2101) [reply R/derive_params_growth_age.R:184]: I originally was trying to do that but couldn't get the logic correctly implemented, if you know how feel free to throw it in!

File: R/derive_params_growth_age.R [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+    ) %>%
+    filter(prev_age <= {{ age }} & {{ age }} < next_age)
+
+  dataset_final <- dataset
+
+  # create separate records objects as appropriate depending if user specific sds and/or pctl
+  if (!is_empty(set_values_to_sds)) {
+    add_sds <- added_records %>%
+      mutate(
+        AVAL = ((VSSTRESN / M)^L - 1) / (L * S),
```
[the remark this answers is not in the document]
Comment (zdz2101) [reply R/derive_params_growth_age.R:221]: it sounds like we need an extra function argument like `analysis_var = VSSTRESN` or `analysis_var = AVAL`

**Comment by zdz2101** [comment]
@rossfarrugia if you want to merge this soon, we can allow others to work in parallel with a "working" function and thoroughly get test/edge cases later as a separate task/issue

@kabis-ops I've addressed or added and emoji to most of the comments thus far, if you wanna take another look please do

File: R/derive_params_growth_age.R (hunk 1 above)
Comment (rossfarrugia) [reply R/derive_params_growth_age.R]: @bundfussr would you be able to advise Zelos on the code to use here please?

File: man/roxygen/meta.R [hunk 14]
```diff
@@ -1,6 +1,7 @@
 list(
   rd_family_title = list(
     metadata = "Metadata",
-    datasets = "Datasets"
+    datasets = "Datasets",
+    der_prm_bds_vs = "der_prm_bds_vs"
```
Comment (rossfarrugia) [inline man/roxygen/meta.R:5]:
```suggestion
    der_prm_bds_vs = "Vital Signs Functions for adding Parameters/Records"
```

File: R/derive_params_growth_age.R [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+                                     age,
+                                     age_unit,
+                                     meta_criteria,
+                                     parameter,
+                                     bmi_cdc_correction = FALSE,
+                                     set_values_to_sds = NULL,
+                                     set_values_to_pctl = NULL) {
+  # Apply assertions to each argument to ensure each object is appropriate class
+  # need to add assertion for dataset that checks columns in sex, age, age_unit are present
+  assert_data_frame(meta_criteria, required_vars = exprs(SEX, AGE, AGEU, L, M, S))
```
[the remark this answers is not in the document]
Comment (rossfarrugia) [reply R/derive_params_growth_age.R:192]: I trust your call here Zelos, but whatever we don't check via assertions should be documented in the function user guide as assumptions so that the user is clear of all expectations

File: R/derive_params_growth_age.R (hunk 13 above)
[the remark this answers is not in the document]
Comment (rossfarrugia) [reply R/derive_params_growth_age.R]: to save another argument could we just always use AVAL here from the input and explain this in the documentation? i was comparing against the approach in https://pharmaverse.github.io/admiral/reference/derive_param_bmi.html and it just says: 

image

File: R/derive_params_growth_age.R (hunk 1 above)
Comment (bundfussr) [reply R/derive_params_growth_age.R]: It seems to be resolved.

**Comment by rossfarrugia** [comment]
@zdz2101 i added comments where i think my input was needed, but shout if any others you want me to look at. 

thinking about your comment above i would keep this PR open and let the review continue here for now until all these comments are closed out rather than merge too early, as me, @Lina2689 and @Fanny-Gautier could just copy your function code from this branch and run in ours if we want to try out the latest for the template or vignette PRs.

then if any of us spot further comments we can add back here, and you have it all controlled in one place rather than having to make subsequent issues/PRs.

i agree we could then make another issue for building out more unit tests later if you prefer, as it gives an opportunity for others in our team to jump in. whoever takes this we should provide the useful doc David sent of tips around testing.

File: R/derive_params_growth_age.R (hunk 1 above)
Comment (bundfussr) [reply R/derive_params_growth_age.R]: By the way, `!!` could be used instead of `{{ }}` when referring to arguments like `sex`, `age`, ... This would be more in line with the other admiral code.

Comment (rossfarrugia) [reply R/derive_params_growth_age.R]: thanks Stefan! sorry, missed that this one was an old comment...

File: R/derive_params_growth_age.R [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+
+  if (is.null(set_values_to_sds) && is.null(set_values_to_pctl)) {
+    abort("One of `set_values_to_sds`/`set_values_to_pctl` has to be specified.")
+  }
+
+  # create a unified join naming convention, hard to figure out in by argument
+  dataset <- dataset %>%
+    mutate(
+      sex_join := {{ sex }},
+      ageu_join := {{ age_unit }}
```
Comment (bundfussr) [inline R/derive_params_growth_age.R:209]: I think these variables should be created with [`get_new_tmp_var()`](https://pharmaverse.github.io/admiraldev/dev/reference/get_new_tmp_var.html) (or just use the specified variables).

File: R/derive_params_growth_age.R [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+#' @examples
+#' library(dplyr)
+#' library(lubridate)
+#' library(rlang)
+#' library(admiral)
+#'
+#' advs <- dm_peds %>%
+#'   select(USUBJID, BRTHDTC, SEX) %>%
+#'   right_join(., vs_peds, by = "USUBJID") %>%
+#'   filter(USUBJID != "PEDS-1010") %>%
```
Comment (rossfarrugia) [inline R/derive_params_growth_age.R:102]: same comments as for the by height function - the test data has changed here now to make patients consistent with pharmaversesdtm so no "PEDS-1010" for example.

**Review by rossfarrugia — APPROVED** [review]
